### PR TITLE
fix the issue of missing set fsGroup for EphemeralStorage

### PR DESCRIPTION
**Pull request by lifupan**
For k8s emptyDir volume, a specific fsGroup would
be set for it, thus guest should get this fsGroup
from runtime and set it properly on the EphemeralStorage
volume in guest.
    
Fixes: #1580

### CODE REVIEW

**Comment by wainersm** [comment]
Hi @lifupan , this fix made my reproducer to pass (https://github.com/kata-containers/tests/pull/3421) and all the k8s e2e tests except for the following:

```
[sig-storage] EmptyDir volumes when FSGroup is specified [LinuxOnly] [NodeFeature:FSGroup] nonexistent volume subPath should have the correct mode and owner using FSGroup [Suite:openshift/conformance/parallel] [Suite:k8s]
```
I only glanced at the test log at  https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/17803/rehearse-17803-periodic-ci-kata-containers-kata-containers-main-e2e-tests-check-fix/1383172767557357568 ; the problem seems to be that the mount point type was expected to be "tmpfs" but actually it got the number "1702057286".

The test fail message:
```
fail [k8s.io/kubernetes@v1.20.0/test/e2e/framework/util.go:738]: Unexpected error:     <*errors.errorString \| 0xc0022f9de0>: {         s: "expected \"mount type of \\\"/test-volume\\\": tmpfs\" in container output: Expected\n    <string>: mount type of \"/test-volume\": 1702057286\n    mode of file \"/test-volume\": dgtrwxrwxrwx\n    perms of file \"/test-volume\": -rwxrwxrwx\n    owner UID of \"/test-volume\": 0\n    owner GID of \"/test-volume\": 123\n    \nto contain substring\n    <string>: mount type of \"/test-volume\": tmpfs",     }     expected "mount type of \"/test-volume\": tmpfs" in container output: Expected         <string>: mount type of "/test-volume": 1702057286         mode of file "/test-volume": dgtrwxrwxrwx         perms of file "/test-volume": -rwxrwxrwx         owner UID of "/test-volume": 0         owner GID of "/test-volume": 123              to contain substring         <string>: mount type of "/test-volume": tmpfs occurred
```

**Comment by lifupan** [comment]
> Hi @lifupan , this fix made my reproducer to pass ([kata-containers/tests#3421](https://github.com/kata-containers/tests/pull/3421)) and all the k8s e2e tests except for the following:
> 
> ```
> [sig-storage] EmptyDir volumes when FSGroup is specified [LinuxOnly] [NodeFeature:FSGroup] nonexistent volume subPath should have the correct mode and owner using FSGroup [Suite:openshift/conformance/parallel] [Suite:k8s]
> ```
> 
> I only glanced at the test log at https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/17803/rehearse-17803-periodic-ci-kata-containers-kata-containers-main-e2e-tests-check-fix/1383172767557357568 ; the problem seems to be that the mount point type was expected to be "tmpfs" but actually it got the number "1702057286".
> 
> The test fail message:
> 
> ```
> fail [k8s.io/kubernetes@v1.20.0/test/e2e/framework/util.go:738]: Unexpected error:     <*errors.errorString \| 0xc0022f9de0>: {         s: "expected \"mount type of \\\"/test-volume\\\": tmpfs\" in container output: Expected\n    <string>: mount type of \"/test-volume\": 1702057286\n    mode of file \"/test-volume\": dgtrwxrwxrwx\n    perms of file \"/test-volume\": -rwxrwxrwx\n    owner UID of \"/test-volume\": 0\n    owner GID of \"/test-volume\": 123\n    \nto contain substring\n    <string>: mount type of \"/test-volume\": tmpfs",     }     expected "mount type of \"/test-volume\": tmpfs" in container output: Expected         <string>: mount type of "/test-volume": 1702057286         mode of file "/test-volume": dgtrwxrwxrwx         perms of file "/test-volume": -rwxrwxrwx         owner UID of "/test-volume": 0         owner GID of "/test-volume": 123              to contain substring         <string>: mount type of "/test-volume": tmpfs occurred
> ```

Hi @wainersm 

I'll have a look on it later.

**Comment by lifupan** [comment]
> Hi @lifupan , this fix made my reproducer to pass ([kata-containers/tests#3421](https://github.com/kata-containers/tests/pull/3421)) and all the k8s e2e tests except for the following:
> 
> ```
> [sig-storage] EmptyDir volumes when FSGroup is specified [LinuxOnly] [NodeFeature:FSGroup] nonexistent volume subPath should have the correct mode and owner using FSGroup [Suite:openshift/conformance/parallel] [Suite:k8s]
> ```
> 
> I only glanced at the test log at https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/17803/rehearse-17803-periodic-ci-kata-containers-kata-containers-main-e2e-tests-check-fix/1383172767557357568 ; the problem seems to be that the mount point type was expected to be "tmpfs" but actually it got the number "1702057286".
> 
> The test fail message:
> 
> ```
> fail [k8s.io/kubernetes@v1.20.0/test/e2e/framework/util.go:738]: Unexpected error:     <*errors.errorString \| 0xc0022f9de0>: {         s: "expected \"mount type of \\\"/test-volume\\\": tmpfs\" in container output: Expected\n    <string>: mount type of \"/test-volume\": 1702057286\n    mode of file \"/test-volume\": dgtrwxrwxrwx\n    perms of file \"/test-volume\": -rwxrwxrwx\n    owner UID of \"/test-volume\": 0\n    owner GID of \"/test-volume\": 123\n    \nto contain substring\n    <string>: mount type of \"/test-volume\": tmpfs",     }     expected "mount type of \"/test-volume\": tmpfs" in container output: Expected         <string>: mount type of "/test-volume": 1702057286         mode of file "/test-volume": dgtrwxrwxrwx         perms of file "/test-volume": -rwxrwxrwx         owner UID of "/test-volume": 0         owner GID of "/test-volume": 123              to contain substring         <string>: mount type of "/test-volume": tmpfs occurred
> ```

Hi @wainersm 

Can you give me a specific steps on how to reproduce this issue, such as a pod yaml and the instructions of how to get the  expected volume info?

**Comment by fidencio** [comment]
> Hi @wainersm
> 
> Can you give me a specific steps on how to reproduce this issue, such as a pod yaml and the instructions of how to get the expected volume info?

Not @wainersm here, but I think this is the test Wainer is using: https://github.com/kata-containers/tests/pull/3421/files

**Comment by wainersm** [comment]
> > Hi @wainersm
> > Can you give me a specific steps on how to reproduce this issue, such as a pod yaml and the instructions of how to get the expected volume info?
> 
> Not @wainersm here, but I think this is the test Wainer is using: https://github.com/kata-containers/tests/pull/3421/files

The test above is passing with current fix. I am working to have a reproducer for the other case.

**Comment by wainersm** [comment]
@lifupan @fidencio ,  I could reproduce the error.

On logs of *mounttest-container-4* and *mounttest-container* the expected mount type should be `tmpfs`. Both have the emptyDir mounted on host filesystem. The other containers are backing the emptyDir in memory, and the mount type is correct.

Another thing that I realized, although I am not sure it is the cause of k8s e2e test failure, is that on both *mounttest-container-3* and *mounttest-container-4* the UID is `0` rather than `1001`. It likely should be fixed too.

```bash
[root@xxxx kubernetes]# cat runtimeclass_workloads/pod-empty-dir-fsgroup.yaml
#
# Copyright (c) 2021 Red Hat, Inc.
#
# SPDX-License-Identifier: Apache-2.0
#
apiVersion: v1
kind: Pod
metadata:
  name: sharevol-kata
spec:
  runtimeClassName: kata
  restartPolicy: Never
  securityContext:
    runAsUser: 1001
    fsGroup: 123
  containers:
  - name: mounttest-container
    image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
    args:
      - mounttest
      - --fs_type=/test-volume
      - --new_file_0660=/test-volume/test-file
      - --file_perm=/test-volume/test-file
      - --file_owner=/test-volume/test-file
    volumeMounts:
      - name: emptydir-volume
        mountPath: /test-volume
  - name: mounttest-container-2
    image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
    args:
      - mounttest
      - --fs_type=/test-volume-2
      - --new_file_0660=/test-volume-2/test-file
      - --file_perm=/test-volume-2/test-file
      - --file_owner=/test-volume-2/test-file
    volumeMounts:
      - name: mem-emptydir-volume
        mountPath: /test-volume-2
  - name: mounttest-container-3
    image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
    args:
      - mounttest
      - --fs_type=/test-volume-3
      - --file_perm=/test-volume-3
      - --file_owner=/test-volume-3
    volumeMounts:
      - name: mem-emptydir-volume
        mountPath: /test-volume-3
  - name: mounttest-container-4
    image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
    args:
      - mounttest
      - --fs_type=/test-volume-4
      - --file_perm=/test-volume-4
      - --file_owner=/test-volume-4
    volumeMounts:
      - name: emptydir-volume
        mountPath: /test-volume-4
  volumes:
  - name: emptydir-volume
    emptyDir: {}
  - name: mem-emptydir-volume
    emptyDir:
      medium: Memory
[root@xxxx kubernetes]# kubectl create -f runtimeclass_workloads/pod-empty-dir-fsgroup.yaml 
pod/sharevol-kata created
[root@xxxx kubernetes]# kubectl get pods
NAME            READY     STATUS      RESTARTS   AGE
sharevol-kata   0/4       Completed   0          23s
[root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-4
mount type of "/test-volume-4": 1702057286
perms of file "/test-volume-4": -rwxrwxrwx
owner UID of "/test-volume-4": 0
owner GID of "/test-volume-4": 123
[root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-3
mount type of "/test-volume-3": tmpfs
perms of file "/test-volume-3": -rwxrwxrwx
owner UID of "/test-volume-3": 0
owner GID of "/test-volume-3": 123
[root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-2
mount type of "/test-volume-2": tmpfs
content of file "/test-volume-2/test-file": mount-tester new file

perms of file "/test-volume-2/test-file": -rw-rw----
owner UID of "/test-volume-2/test-file": 1001
owner GID of "/test-volume-2/test-file": 123
[root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container
mount type of "/test-volume": 1702057286
content of file "/test-volume/test-file": mount-tester new file

perms of file "/test-volume/test-file": -rw-rw----
owner UID of "/test-volume/test-file": 1001
owner GID of "/test-volume/test-file": 123
```

**Comment by wainersm** [comment]
wait... I check the logs for non-kata containers and got a similar result as I posted before. This is weird.

```
[root@xxxx kubernetes]# kubectl logs sharevol-non-kata mounttest-container-4
mount type of "/test-volume-4": 1481003842
perms of file "/test-volume-4": -rwxrwxrwx
owner UID of "/test-volume-4": 0
owner GID of "/test-volume-4": 123
[root@xxxx kubernetes]# kubectl logs sharevol-non-kata mounttest-container-3
mount type of "/test-volume-3": tmpfs
perms of file "/test-volume-3": -rwxrwxrwx
owner UID of "/test-volume-3": 0
owner GID of "/test-volume-3": 123
[root@xxxx kubernetes]# kubectl logs sharevol-non-kata mounttest-container-2
mount type of "/test-volume-2": tmpfs
content of file "/test-volume-2/test-file": mount-tester new file

perms of file "/test-volume-2/test-file": -rw-rw----
owner UID of "/test-volume-2/test-file": 1001
owner GID of "/test-volume-2/test-file": 123
[root@xxxx kubernetes]# kubectl logs sharevol-non-kata mounttest-container
mount type of "/test-volume": 1481003842
content of file "/test-volume/test-file": mount-tester new file

perms of file "/test-volume/test-file": -rw-rw----
owner UID of "/test-volume/test-file": 1001
owner GID of "/test-volume/test-file": 123
```

**Comment by lifupan** [comment]
> @lifupan @fidencio , I could reproduce the error.
> 
> On logs of _mounttest-container-4_ and _mounttest-container_ the expected mount type should be `tmpfs`. Both have the emptyDir mounted on host filesystem. The other containers are backing the emptyDir in memory, and the mount type is correct.
> 
> Another thing that I realized, although I am not sure it is the cause of k8s e2e test failure, is that on both _mounttest-container-3_ and _mounttest-container-4_ the UID is `0` rather than `1001`. It likely should be fixed too.
> 
> ```shell
> [root@xxxx kubernetes]# cat runtimeclass_workloads/pod-empty-dir-fsgroup.yaml
> #
> # Copyright (c) 2021 Red Hat, Inc.
> #
> # SPDX-License-Identifier: Apache-2.0
> #
> apiVersion: v1
> kind: Pod
> metadata:
>   name: sharevol-kata
> spec:
>   runtimeClassName: kata
>   restartPolicy: Never
>   securityContext:
>     runAsUser: 1001
>     fsGroup: 123
>   containers:
>   - name: mounttest-container
>     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
>     args:
>       - mounttest
>       - --fs_type=/test-volume
>       - --new_file_0660=/test-volume/test-file
>       - --file_perm=/test-volume/test-file
>       - --file_owner=/test-volume/test-file
>     volumeMounts:
>       - name: emptydir-volume
>         mountPath: /test-volume
>   - name: mounttest-container-2
>     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
>     args:
>       - mounttest
>       - --fs_type=/test-volume-2
>       - --new_file_0660=/test-volume-2/test-file
>       - --file_perm=/test-volume-2/test-file
>       - --file_owner=/test-volume-2/test-file
>     volumeMounts:
>       - name: mem-emptydir-volume
>         mountPath: /test-volume-2
>   - name: mounttest-container-3
>     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
>     args:
>       - mounttest
>       - --fs_type=/test-volume-3
>       - --file_perm=/test-volume-3
>       - --file_owner=/test-volume-3
>     volumeMounts:
>       - name: mem-emptydir-volume
>         mountPath: /test-volume-3
>   - name: mounttest-container-4
>     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
>     args:
>       - mounttest
>       - --fs_type=/test-volume-4
>       - --file_perm=/test-volume-4
>       - --file_owner=/test-volume-4
>     volumeMounts:
>       - name: emptydir-volume
>         mountPath: /test-volume-4
>   volumes:
>   - name: emptydir-volume
>     emptyDir: {}
>   - name: mem-emptydir-volume
>     emptyDir:
>       medium: Memory
> [root@xxxx kubernetes]# kubectl create -f runtimeclass_workloads/pod-empty-dir-fsgroup.yaml 
> pod/sharevol-kata created
> [root@xxxx kubernetes]# kubectl get pods
> NAME            READY     STATUS      RESTARTS   AGE
> sharevol-kata   0/4       Completed   0          23s
> [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-4
> mount type of "/test-volume-4": 1702057286
> perms of file "/test-volume-4": -rwxrwxrwx
> owner UID of "/test-volume-4": 0
> owner GID of "/test-volume-4": 123
> [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-3
> mount type of "/test-volume-3": tmpfs
> perms of file "/test-volume-3": -rwxrwxrwx
> owner UID of "/test-volume-3": 0
> owner GID of "/test-volume-3": 123
> [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-2
> mount type of "/test-volume-2": tmpfs
> content of file "/test-volume-2/test-file": mount-tester new file
> 
> perms of file "/test-volume-2/test-file": -rw-rw----
> owner UID of "/test-volume-2/test-file": 1001
> owner GID of "/test-volume-2/test-file": 123
> [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container
> mount type of "/test-volume": 1702057286
> content of file "/test-volume/test-file": mount-tester new file
> 
> perms of file "/test-volume/test-file": -rw-rw----
> owner UID of "/test-volume/test-file": 1001
> owner GID of "/test-volume/test-file": 123
> ```

Hi @wainersm 

For the emptyDir volume without the "Memory" medium backend, it should be just a host directory and its mount type shouldn't be tmpfs, and the expected tmpfs mount type only occur when the backend media is memory.   Can you double check your testcase?

**Comment by wainersm** [comment]
> > @lifupan @fidencio , I could reproduce the error.
> > On logs of _mounttest-container-4_ and _mounttest-container_ the expected mount type should be `tmpfs`. Both have the emptyDir mounted on host filesystem. The other containers are backing the emptyDir in memory, and the mount type is correct.
> > Another thing that I realized, although I am not sure it is the cause of k8s e2e test failure, is that on both _mounttest-container-3_ and _mounttest-container-4_ the UID is `0` rather than `1001`. It likely should be fixed too.
> > ```shell
> > [root@xxxx kubernetes]# cat runtimeclass_workloads/pod-empty-dir-fsgroup.yaml
> > #
> > # Copyright (c) 2021 Red Hat, Inc.
> > #
> > # SPDX-License-Identifier: Apache-2.0
> > #
> > apiVersion: v1
> > kind: Pod
> > metadata:
> >   name: sharevol-kata
> > spec:
> >   runtimeClassName: kata
> >   restartPolicy: Never
> >   securityContext:
> >     runAsUser: 1001
> >     fsGroup: 123
> >   containers:
> >   - name: mounttest-container
> >     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
> >     args:
> >       - mounttest
> >       - --fs_type=/test-volume
> >       - --new_file_0660=/test-volume/test-file
> >       - --file_perm=/test-volume/test-file
> >       - --file_owner=/test-volume/test-file
> >     volumeMounts:
> >       - name: emptydir-volume
> >         mountPath: /test-volume
> >   - name: mounttest-container-2
> >     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
> >     args:
> >       - mounttest
> >       - --fs_type=/test-volume-2
> >       - --new_file_0660=/test-volume-2/test-file
> >       - --file_perm=/test-volume-2/test-file
> >       - --file_owner=/test-volume-2/test-file
> >     volumeMounts:
> >       - name: mem-emptydir-volume
> >         mountPath: /test-volume-2
> >   - name: mounttest-container-3
> >     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
> >     args:
> >       - mounttest
> >       - --fs_type=/test-volume-3
> >       - --file_perm=/test-volume-3
> >       - --file_owner=/test-volume-3
> >     volumeMounts:
> >       - name: mem-emptydir-volume
> >         mountPath: /test-volume-3
> >   - name: mounttest-container-4
> >     image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
> >     args:
> >       - mounttest
> >       - --fs_type=/test-volume-4
> >       - --file_perm=/test-volume-4
> >       - --file_owner=/test-volume-4
> >     volumeMounts:
> >       - name: emptydir-volume
> >         mountPath: /test-volume-4
> >   volumes:
> >   - name: emptydir-volume
> >     emptyDir: {}
> >   - name: mem-emptydir-volume
> >     emptyDir:
> >       medium: Memory
> > [root@xxxx kubernetes]# kubectl create -f runtimeclass_workloads/pod-empty-dir-fsgroup.yaml 
> > pod/sharevol-kata created
> > [root@xxxx kubernetes]# kubectl get pods
> > NAME            READY     STATUS      RESTARTS   AGE
> > sharevol-kata   0/4       Completed   0          23s
> > [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-4
> > mount type of "/test-volume-4": 1702057286
> > perms of file "/test-volume-4": -rwxrwxrwx
> > owner UID of "/test-volume-4": 0
> > owner GID of "/test-volume-4": 123
> > [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-3
> > mount type of "/test-volume-3": tmpfs
> > perms of file "/test-volume-3": -rwxrwxrwx
> > owner UID of "/test-volume-3": 0
> > owner GID of "/test-volume-3": 123
> > [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container-2
> > mount type of "/test-volume-2": tmpfs
> > content of file "/test-volume-2/test-file": mount-tester new file
> > 
> > perms of file "/test-volume-2/test-file": -rw-rw----
> > owner UID of "/test-volume-2/test-file": 1001
> > owner GID of "/test-volume-2/test-file": 123
> > [root@xxxx kubernetes]# kubectl logs sharevol-kata mounttest-container
> > mount type of "/test-volume": 1702057286
> > content of file "/test-volume/test-file": mount-tester new file
> > 
> > perms of file "/test-volume/test-file": -rw-rw----
> > owner UID of "/test-volume/test-file": 1001
> > owner GID of "/test-volume/test-file": 123
> > ```
> 
> Hi @wainersm
> 
> For the emptyDir volume without the "Memory" medium backend, it should be just a host directory and its mount type shouldn't be tmpfs, and the expected tmpfs mount type only occur when the backend media is memory. Can you double check your testcase?

Yes, @lifupan , I will have a deeper look at the original e2e test to come up with a reproducer.

Well, for now what we know is that this fix will make my original reproducer (in https://github.com/kata-containers/tests/pull/3421) to pass, and some e2e tests are also passing. So maybe we should go ahead and merge this (once approved by reviewers, of course), then I can open another issue later when I have the new reproducer in place. Makes sense?

**Comment by wainersm** [comment]
@lifupan , ok, now I can reproduce the problem, it is related with [volumes subPath](https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath). If a *subPath* is defined then the mount type is not *tmpfs*. My feeling is that in this case the *emptDir* is backed on host filesystem (rather than on memory, as specified on the pod configuration).

It seems not related with *fsGroup* as I could reproduce the error even without setting it. So, as I stated in a previous comment, if you prefer we can merge this fix and work on another PR.

Following the results of my experiments:

```
[root@xxxx kubernetes]# cat runtimeclass_workloads/pod-empty-dir-fsgroup.yaml
#
# Copyright (c) 2021 Red Hat, Inc.
#
# SPDX-License-Identifier: Apache-2.0
#
apiVersion: v1
kind: Pod
metadata:
  name: sharevol-kata
spec:
  runtimeClassName: kata
  restartPolicy: Never
  securityContext:
    runAsUser: 1001
    fsGroup: 123
  containers:
  - name: mounttest-container-3
    image: "k8s.gcr.io/e2e-test-images/agnhost:2.21"
    args:
      - mounttest
      - --fs_type=/test-volume-3
      - --file_perm=/test-volume-3
      - --file_owner=/test-volume-3
      - --file_mode=/test-volume-3
    volumeMounts:
      - name: mem-emptydir-volume
        mountPath: /test-volume-3
        subPath: test-sub
  volumes:
  - name: mem-emptydir-volume
    emptyDir:
      medium: Memory
[root@xxxx kubernetes]# kubectl create -f runtimeclass_workloads/pod-empty-dir-fsgroup.yaml 
pod/sharevol-kata created
[root@xxxx kubernetes]# kubectl logs sharevol-kata
mount type of "/test-volume-3": 1702057286
mode of file "/test-volume-3": dgtrwxrwxrwx
perms of file "/test-volume-3": -rwxrwxrwx
owner UID of "/test-volume-3": 0
owner GID of "/test-volume-3": 123
```
And this is the output which I got when using `runc`:

```
[root@xxxx kubernetes]# kubectl logs sharevol-non-kata
mount type of "/test-volume-3": tmpfs
mode of file "/test-volume-3": dgtrwxrwxrwx
perms of file "/test-volume-3": -rwxrwxrwx
owner UID of "/test-volume-3": 0
owner GID of "/test-volume-3": 123
```

In a previous comment I mentioned about UID set to zero and it is correct; so ignore my comment.

**Comment by lifupan** [comment]
> @lifupan , ok, now I can reproduce the problem, it is related with [volumes subPath](https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath). If a _subPath_ is defined then the mount type is not _tmpfs_. My feeling is that in this case the _emptDir_ is backed on host filesystem (rather than on memory, as specified on the pod configuration).
> 
> It seems not related with _fsGroup_ as I could reproduce the error even without setting it. So, as I stated in a previous comment, if you prefer we can merge this fix and work on another PR.
> 

Hi @wainersm 
It's reasonable, and please go ahead and file another issue and I'll push to merge this PR.

File: src/agent/src/mount.rs
Comment (liubin) [inline src/agent/src/mount.rs:253]: Insert a space after `//`

Comment (lifupan) [reply src/agent/src/mount.rs]: Hi @liubin 

Done.

**Review by liubin — APPROVED** [review]

File: src/runtime/virtcontainers/kata_agent_test.go
Comment (bergwolf) [inline src/runtime/virtcontainers/kata_agent_test.go:195]: assert error here?

**Review by bergwolf — COMMENTED** [review]
Just one nit. Otherwise lgtm!

File: src/runtime/virtcontainers/kata_agent_test.go
Comment (lifupan) [reply src/runtime/virtcontainers/kata_agent_test.go]: Hi @bergwolf 

Updated.

**Review by bergwolf — APPROVED** [review]

**Comment by fidencio** [comment]
@lifupan, static check is failing for 1.13:
```
src/runtime/virtcontainers/kata_agent_test.go:196:13: cannot use err (variable of type error) as assert.TestingT value in argument to assert.Nil: missing method Errorf (typecheck)
	assert.Nil(err)
	           ^
Error: Process completed with exit code 1.
```

**Pull request merged**
